donation-tracker/src/utils/dataStore.ts

```diff
@@ -1,316 +1,486 @@
+    });
+    return result;
+  }
+
+  /**
+   * Internal function to validate the column mapping (were all expected columns found)
+   * @param mapping The mapping that was generated from the createColumnMapping function
+   * @returns true if everything is looking good, false if a column is not found (or not available for all languages)
+   */
+  private validateColumnMapping(mapping: ColumnMapping): boolean {
+    this.debugLog('DataStore: Validating column mapping');
+    let result = true;
+    // simple value columns
+    if (mapping.campaignId === -1) {
+      this.errorLog('DataStore: Column Mapping Validation failed -> campaign ID column not found');
+      result = false;
+    }
+    if (mapping.donated === -1) {
+      this.errorLog('DataStore: Column Mapping Validation failed -> donated amount column not found');
+      result = false;
+    }
+    if (mapping.need === -1) {
+      this.errorLog('DataStore: Column Mapping Validation failed -> needed amount column not found');
+      result = false;
+    }
+    if (mapping.lastChange === -1) {
+      this.errorLog('DataStore: Column Mapping Validation failed -> last data update column not found');
+      result = false;
+    }
+    // localized value columns
+    mapping.languages.forEach(value => {
+      if (mapping.item[value] === -1 || mapping.item[value] === undefined) {
+        this.errorLog(`DataStore: Column Mapping Validation failed -> item name column for language '${value}' not found`);
+        result = false;
+      }
+      if (mapping.unit[value] === -1 || mapping.item[value] === undefined) {
+        this.errorLog(`DataStore: Column Mapping Validation failed -> unit column for language '${value}' not found`);
+        result = false;
+      }
+      if (mapping.form[value] === -1 || mapping.item[value] === undefined) {
+        this.errorLog(`DataStore: Column Mapping Validation failed -> form URL column for language '${value}' not found`);
+        result = false;
+      }
+    });
+    return result;
+  }
+
+  /**
+   * Internal function to load / refresh the local data
+   */
+  private async refreshData() {
+    this.isRefreshing = true;
+    this.debugLog('DataStore: Data refresh started');
+    let data = await fetch(this.dataSourceUrl, { cache: 'no-store' });
+    const content = await data.text();
+    const rows = content.split(`\n`);
+    const tmpParsedData: DonationList = {
+      version: this.dataVersion,
+      timeStamp: new Date(1970, 0, 1),
+      requestTime: new Date(),
+      data: [],
+    };
+    if (rows.length > 0) {
+      const mapping = this.createColumnMapping(rows[0].trim());
+      if (!this.validateColumnMapping(mapping)) {
+        this.isRefreshing = false;
+        this.registerDataUpdate();
+        return;
+      }
+
+      this.debugLog('DataStore: Data received');
+      // extract line with change timestamp
+      let timestampRow = this.parseCsvLine(rows[1].trim());
+      if (timestampRow.length >= mapping.lastChange + 1) {
```

Comment:
I wonder if this check is still needed since we now have somewhat "absolute positioning" of columns by full header detection.

Reply:
I'll keep this as an empty column (someone accidentally deleting the date) would result in a "out of bounds" - as the RegEx row splitter can not really deal with empty columns (shortcoming to handle later; can be removed once this flaw is dealt with)
But valid point